### PR TITLE
Create example hooks section in documentation

### PR DESCRIPTION
**Pull request by thomasnares**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 8.x
| Description?  | Create a section for example hooks implementations, to be completed in the next few months.

### CODE REVIEW

**Comment by thomasnares** [comment]
@PululuK here it is :) I authored you for the example

**Comment by thomasnares** [comment]
Related to #1543

**Review by nicosomb — APPROVED** [review]

File: modules/sample-modules/_index.md [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
+
+The PrestaShop Community made some example modules and how to guides to help you understand how to implement hooks, admin grids, manage entities... and a lot more.
 
 {{% children %}} 
 
+## Our example modules repository
+
 * [See all our example modules (GitHub)](https://github.com/PrestaShop/example-modules)
+
+This repository hosts example modules built and maintained by PrestaShop.
```
Comment (kpodemski) [inline modules/sample-modules/_index.md]:
```suggestion
This repository hosts example modules built and maintained by project members and the community.
```

File: modules/sample-modules/_index.md [hunk 2]
```diff
@@ -1,10 +1,46 @@
...
+These modules demonstrate useful usecases for developers willing to customize the software.
```
Comment (kpodemski) [inline modules/sample-modules/_index.md]:
```suggestion
These modules demonstrate good use cases for developers willing to customize the software.
```

File: modules/sample-modules/example-hooks/_index.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+---
+title: Hooks examples
+weight: 2
+---
+
+# Hooks implementation examples
+
+Hooks is one of the most powerful tools to extend or modify PrestaShop with modules. 
+
+For a complete introduction about hooks, please read [Hooks]({{< relref "/8/modules/concepts/hooks/" >}})
```
Comment (kpodemski) [inline modules/sample-modules/example-hooks/_index.md]:
```suggestion
For a complete introduction to hooks, please read [Hooks]({{< relref "/8/modules/concepts/hooks/" >}})
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    array(
+      'cart' => (object) Cart,
+      'order' => (object) Order,
+      'customer' => (object) Customer,
+      'currency' => (object) Currency,
+      'orderStatus' => (object) OrderState
+    );
+    ```
+    
+A classic use-case for this hook could be :
```
Comment (kpodemski) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
A classic use case for this hook could be: 
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            
+            //TODO : of course don't forget to log if something fails here :)
+                }
+            }
+        }
+    }
+
+    protected function customerAlreadyRewarded(int $idCustomer): bool
+    {
+        //TODO : check if customer already rewarded
```
Comment (kpodemski) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        //TODO: check if customer already rewarded
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    protected setAlreadyRewarded(): void
+    {
+        //TODO: set customer was rewarded
+    }
+
+    protected function getConfuredOrdersStatesIds(): array
+    {
+        //TODO : return array with configured states ids in your module
```
Comment (kpodemski) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        //TODO: return array with configured states ids in your module
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    protected function getCustomerValidOrdersNbr(int $idCustomer): int
+    {
+        //TODO : return number of total order valid by customer
```
Comment (kpodemski) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        //TODO: return number of total order valid by customer
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    protected function getRequiredNbrOfTheOrderToReward(): int
+    {
+        //TODO : return configured number of orders required to reward the customer
```
Comment (kpodemski) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        //TODO: return configured number of orders required to reward the customer
```

File: modules/sample-modules/_index.md [hunk 9]
```diff
@@ -1,10 +1,46 @@
 ---
-title: Sample modules
+title: Sample modules and how to's
 weight: 80
 ---
 
-# Sample modules
+# Sample modules and how to guides
+
+The PrestaShop Community made some example modules and how to guides to help you understand how to implement hooks, admin grids, manage entities... and a lot more.
```
Comment (eternoendless) [inline modules/sample-modules/_index.md]:
```suggestion
The PrestaShop Community has made example modules and how-to guides to help you understand how to implement hooks, admin grids, manage entities... and a lot more.
```

File: modules/sample-modules/_index.md [hunk 10]
```diff
@@ -1,10 +1,46 @@
...
 
 {{% children %}} 
 
+## Our example modules repository
```
Comment (eternoendless) [inline modules/sample-modules/_index.md]:
```suggestion
## Example modules repository
```

File: modules/sample-modules/_index.md [hunk 11]
```diff
@@ -1,10 +1,46 @@
...
+
 * [See all our example modules (GitHub)](https://github.com/PrestaShop/example-modules)
+
+This repository hosts example modules built and maintained by PrestaShop.
+These modules demonstrate useful usecases for developers willing to customize the software.
+
+### Extending Symfony pages
+
+- [Demo Extending a Symfony Form - 1](https://github.com/PrestaShop/example-modules/tree/master/demoextendsymfonyform1) - how to insert an input inside a Symfony form
```
Comment (eternoendless) [inline modules/sample-modules/_index.md]: Consider putting the right text inside the anchor, it's better for SEO. Consider the search terms a person would use when googling for this information.
```suggestion
- [Extending a Symfony form - how to insert a field inside an existing form](https://github.com/PrestaShop/example-modules/tree/master/demoextendsymfonyform1)
```

File: modules/sample-modules/example-hooks/_index.md [hunk 12]
```diff
@@ -0,0 +1,12 @@
+---
+title: Hooks examples
```
Comment (eternoendless) [inline modules/sample-modules/example-hooks/_index.md]:
```suggestion
title: Hook examples
```

File: modules/sample-modules/example-hooks/_index.md [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+weight: 2
+---
+
+# Hooks implementation examples
```
Comment (eternoendless) [inline modules/sample-modules/example-hooks/_index.md]:
```suggestion
# Hook implementation examples
```

File: modules/sample-modules/example-hooks/_index.md [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+
+Hooks is one of the most powerful tools to extend or modify PrestaShop with modules.
```
Comment (eternoendless) [inline modules/sample-modules/example-hooks/_index.md]:
```suggestion
Hooks are one of the most powerful tools to extend or modify PrestaShop using modules. 
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
+---
+title: "actionValidateOrder"
```
Comment (eternoendless) [inline modules/sample-modules/example-hooks/actionValidateOrder.md:2]: FYI you don't usually need quotes for strings in yaml unless you're using special characters, integers or `true|false`

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+weight: 3
+---
+
+# Hook example : actionValidateOrder
```
Comment (eternoendless) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]: Remember that in English there is no space before `[:;!?]`
```suggestion
# Hook example: actionValidateOrder
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+      'order' => (object) Order,
+      'customer' => (object) Customer,
+      'currency' => (object) Currency,
+      'orderStatus' => (object) OrderState
+    );
+    ```
+    
+A classic use-case for this hook could be : 
+
+> I want to reward my customers on their _n-th_ order
```
Comment (eternoendless) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]: Avoid using block quotes unless you're actually quoting someone
```suggestion
_I want to reward my customers on their n-th order_
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+            $hasCustomerRequiredNbrOfTheOrderToReward = $this->getCustomerValidOrdersNbr((int) $customerObject->id) == $this->getRequiredNbrOfTheOrderToReward();
+            if($hasConfiguredState && $hasCustomerRequiredNbrOfTheOrderToReward) {
+                $customerReward = $this->createCustomerReward($customerObject, $orderObject);
+                if(Validate::isLoadedObject($customerReward)) {
+                        $this->setAlreadyRewarded($customerObject);
+                        $this->notifyCustomer($customerObject, $customerReward);
+            
+            //TODO : of course don't forget to log if something fails here :)
+                }
+            }
```
Comment (eternoendless) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]: Fixed some tabbing issues here. Is it really necessary to keep the "TODO"s?
```suggestion
            if ($hasConfiguredState && $hasCustomerRequiredNbrOfTheOrderToReward) {
                $customerReward = $this->createCustomerReward($customerObject, $orderObject);
                if (Validate::isLoadedObject($customerReward)) {
                    $this->setAlreadyRewarded($customerObject);
                    $this->notifyCustomer($customerObject, $customerReward);
            
                    // of course don't forget to log if something fails here :)
                }
            }
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md (hunk 5 above)
Comment (thomasnares) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        // check if customer already rewarded
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    protected function customerAlreadyRewarded(int $idCustomer): bool
+    {
+        //TODO : check if customer already rewarded
+    }
+
+    protected setAlreadyRewarded(): void
+    {
+        //TODO: set customer was rewarded
```
Comment (thomasnares) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        // set customer was rewarded
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md (hunk 6 above)
Comment (thomasnares) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        // return array with configured states ids in your module
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md (hunk 7 above)
Comment (thomasnares) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        // return number of total order valid by customer
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md (hunk 8 above)
Comment (thomasnares) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        // return configured number of orders required to reward the customer
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    protected function getRequiredNbrOfTheOrderToReward(): int
+    {
+        //TODO : return configured number of orders required to reward the customer
+    }
+
+    protected function createCustomerReward(Customer $customer, Order $order): ?CartRule
+    {
+        //TODO: generate customer cart rule (according to the order amount for example)
```
Comment (thomasnares) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        // generate customer cart rule (according to the order amount for example)
```

File: modules/sample-modules/example-hooks/actionValidateOrder.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    protected function notifyCustomer(Customer $customer, CartRule $cartRule): bool
+    {
+        //TODO: notify the customer
```
Comment (thomasnares) [inline modules/sample-modules/example-hooks/actionValidateOrder.md]:
```suggestion
        // notify the customer 
```

**Review by kpodemski — APPROVED** [review]
Thank you @thomasnares good start!

**Comment by PululuK** [comment]
Great @thomasnares  !! Thanks

**Pull request merged**